pkg/apis/kubeone/types.go

```diff
@@ -52,6 +52,23 @@ type KubeOneCluster struct {
 	Credentials map[string]string `json:"credentials,omitempty"`
 }
 
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// KubeOneSecrets hold secrets needed to provision the cluster
+type KubeOneSecrets struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Secrets contains list of secrets to be used by KubeOne instead of
+	// environment variables
+	Secrets map[string]string `json:"secrets"`
```

Comment:
This has both positive and negative sides.

From the positive side, we don't have a very complex type and users can just provide secrets similar to environment variables.

From the negative side, we don't have strong type validation, i.e. everything is a string and we can't validate names out-of-box without some more complex function (e.g. `credentials.ProviderCredentials`).